src/geometry/manifold/diffeomorph.lean

```diff
@@ -39,64 +53,365 @@ variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
   M ≃ₘ^n⟮I, J⟯ N :=
 { times_cont_mdiff_to_fun  := h₂.times_cont_mdiff_to_fun.comp h₁.times_cont_mdiff_to_fun,
   times_cont_mdiff_inv_fun := h₁.times_cont_mdiff_inv_fun.comp h₂.times_cont_mdiff_inv_fun,
-  .. equiv.trans h₁.to_equiv h₂.to_equiv }
+  to_equiv := h₁.to_equiv.trans h₂.to_equiv }
+
+@[simp] lemma trans_refl (h : M ≃ₘ^n⟮I, I'⟯ M') : h.trans (diffeomorph.refl I' M' n) = h :=
+ext $ λ _, rfl
+@[simp] lemma refl_trans (h : M ≃ₘ^n⟮I, I'⟯ M') : (diffeomorph.refl I M n).trans h = h :=
+ext $ λ _, rfl
+@[simp] lemma coe_trans (h₁ : M ≃ₘ^n⟮I, I'⟯ M') (h₂ : M' ≃ₘ^n⟮I', J⟯ N) :
+  ⇑(h₁.trans h₂) = h₂ ∘ h₁ := rfl
 
 /-- Inverse of a diffeomorphism. -/
 protected def symm (h : M ≃ₘ^n⟮I, J⟯ N) : N ≃ₘ^n⟮J, I⟯ M :=
 { times_cont_mdiff_to_fun  := h.times_cont_mdiff_inv_fun,
   times_cont_mdiff_inv_fun := h.times_cont_mdiff_to_fun,
-  .. h.to_equiv.symm }
+  to_equiv := h.to_equiv.symm }
+
+@[simp] lemma apply_symm_apply (h : M ≃ₘ^n⟮I, J⟯ N) (x : N) : h (h.symm x) = x :=
+h.to_equiv.apply_symm_apply x
+@[simp] lemma symm_apply_apply (h : M ≃ₘ^n⟮I, J⟯ N) (x : M) : h.symm (h x) = x :=
+h.to_equiv.symm_apply_apply x
+
+@[simp] lemma symm_refl : (diffeomorph.refl I M n).symm = diffeomorph.refl I M n :=
+ext $ λ _, rfl
+@[simp] lemma trans_symm (h : M ≃ₘ^n⟮I, J⟯ N) : h.trans h.symm = diffeomorph.refl I M n :=
+ext h.symm_apply_apply
+@[simp] lemma symm_trans (h : M ≃ₘ^n⟮I, J⟯ N) : h.symm.trans h = diffeomorph.refl J N n :=
+ext h.apply_symm_apply
+@[simp] lemma symm_trans' (h₁ : M ≃ₘ^n⟮I, I'⟯ M') (h₂ : M' ≃ₘ^n⟮I', J⟯ N) :
+  (h₁.trans h₂).symm = h₂.symm.trans h₁.symm := rfl
+@[simp] lemma to_equiv_symm (h : M ≃ₘ^n⟮I, J⟯ N) : h.symm.to_equiv = h.to_equiv.symm := rfl
+@[simp, mfld_simps] lemma to_equiv_coe_symm (h : M ≃ₘ^n⟮I, J⟯ N) : ⇑h.to_equiv.symm = h.symm := rfl
+
+lemma image_eq_preimage (h : M ≃ₘ^n⟮I, J⟯ N) (s : set M) : h '' s = h.symm ⁻¹' s :=
+h.to_equiv.image_eq_preimage s
+lemma symm_image_eq_preimage (h : M ≃ₘ^n⟮I, J⟯ N) (s : set N) : h.symm '' s = h ⁻¹' s :=
+h.symm.image_eq_preimage s
+
+@[simp] lemma image_symm_image (h : M ≃ₘ^n⟮I, J⟯ N) (s : set N) : h '' (h.symm '' s) = s :=
+h.to_equiv.image_symm_image s
+@[simp] lemma symm_image_image (h : M ≃ₘ^n⟮I, J⟯ N) (s : set M) : h.symm '' (h '' s) = s :=
+h.to_equiv.symm_image_image s
+
+/-- A diffeomorphism is a homeomorphism. -/
+def to_homeomorph (h : M ≃ₘ^n⟮I, J⟯ N) : M ≃ₜ N :=
+⟨h.to_equiv, h.continuous, h.symm.continuous⟩
+
+@[simp] lemma to_homeomorph_to_equiv (h : M ≃ₘ^n⟮I, J⟯ N) :
+  h.to_homeomorph.to_equiv = h.to_equiv :=
+rfl
+@[simp] lemma symm_to_homeomorph (h : M ≃ₘ^n⟮I, J⟯ N) :
+  h.symm.to_homeomorph = h.to_homeomorph.symm :=
+rfl
+
+@[simp] lemma coe_to_homeomorph (h : M ≃ₘ^n⟮I, J⟯ N) : ⇑h.to_homeomorph = h := rfl
+@[simp] lemma coe_to_homeomorph_symm (h : M ≃ₘ^n⟮I, J⟯ N) :
+  ⇑h.to_homeomorph.symm = h.symm := rfl
+
+@[simp] lemma comp_diffeomorph_times_cont_mdiff_within_at_iff {m} (h : M ≃ₘ^n⟮I, J⟯ N) {f : N → M'}
```

Comment:
The name feels strange here. Should it be `times_cont_mdiff_within_at_comp_diffeomorph_iff`? Also, it's a little bit strange to have this in the `diffeomorph` namespace, but I understand you don't want to exit it. If we had lean#550, you could just prefix it with `_root_`...

Same thing for several lemmas below.

Reply:
Renamed.